lib/Dialect/HW/Transforms/LowerHWtoBTOR2.cpp

```diff
@@ -0,0 +1,786 @@
+//===- LowerHWtoBTOR2.cpp - Lowers a hw module to btor2 --------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Converts a hw module to a btor2 format and prints it out
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/Comb/CombDialect.h"
+#include "circt/Dialect/HW/HWModuleGraph.h"
+#include "circt/Dialect/HW/HWPasses.h"
+#include "circt/Dialect/SV/SVDialect.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/SV/SVTypes.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/GraphWriter.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace circt;
+using namespace hw;
+
+// Macros for C-style error handling
+#define NO_LID (-1UL)
+#define NO_WIDTH (-1L)
+
+namespace {
+// The goal here is to traverse the operations in order and convert them one by
+// one into btor2
+struct LowerHWtoBTOR2Pass : public LowerHWtoBTOR2Base<LowerHWtoBTOR2Pass> {
+public:
+  // Executes the pass
+  void runOnOperation() override;
+
+private:
+  // Create a counter that attributes a unique id to each generated btor2 line
+  size_t lid = 1;           // btor2 line identifiers usually start at 1
+  size_t resetLID = NO_LID; // keeps track of the reset's LID
+
+  // Create maps to keep track of lid associations
+  // Proper maps wouldn't work for some reason so I'll use a vector of pairs
+  // instead
+  SmallVector<std::pair<size_t, size_t>>
+      sortToLIDMap; // Keeps track of the ids associated to each declared sort
+  DenseMap<std::pair<int64_t, size_t>, size_t>
+      constToLIDMap; // Keeps track of the
+  DenseMap<Operation *, size_t>
+      opLIDMap; // Connects an operation to it's most recent update line
```

Comment:
You might want to use Value instead of Operation* here, I haven't looked at what you are doing, but that would allow distinguishing between results from ops.  It also let's you handle blockargs which would let you flatten module hierarchies on the fly.

Reply:
That was what I had originally, but I don't always have access to values, so I opted for Operation* (might also simply be something I misunderstood)